src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs

```diff
@@ -83,32 +83,101 @@ internal partial class CodeFixService : ForegroundThreadAffinitizedObject, ICode
+                                               () => GetFirstDiagnosticWithFix(document, severityGroup, range, linkedToken), linkedToken, TaskScheduler.Default)).ToSet();
+
+                var errorDiagnosticTask = AddErrorDiagnosticTask(document, range, diagnosticGroups, taskSet, linkedToken);
 
-                    // REVIEW: 2 possible designs. 
-                    // 1. find the first fix and then return right away. if the lightbulb is actually expanded, find all fixes for the line synchronously. or
-                    // 2. kick off a task that finds all fixes for the given range here but return once we find the first one. 
-                    // at the same time, let the task to run to finish. if the lightbulb is expanded, we just simply use the task to get all fixes.
-                    //
-                    // first approach is simpler, so I will implement that first. if the first approach turns out to be not good enough, then
-                    // I will try the second approach which will be more complex but quicker
-                    var hasFix = await ContainsAnyFix(document, diagnostic, cancellationToken).ConfigureAwait(false);
-                    if (hasFix)
+                var results = new List<DiagnosticData>();
+                try
+                {
+                    while (taskSet.Count > 0)
                     {
-                        return new FirstDiagnosticResult(!fullResult, hasFix, diagnostic);
+                        var completedTask = await Task.WhenAny(taskSet).ConfigureAwait(false);
+                        if (completedTask == errorDiagnosticTask && completedTask.Result != null)
+                        {
+                            // Error task returned a result; just return it.
+                            return CreateResult(fullResult, completedTask.Result);
+                        }
+                        else if (errorDiagnosticTask == null ||
+                            errorDiagnosticTask.IsCompleted && errorDiagnosticTask.Result == null)
+                        {
+                            // Some other task completed and we know there's no fixable error diagnostic.
+                            return CreateResult(fullResult, completedTask.Result);
+                        }
+                        else
+                        {
+                            // We don't know if we're going to show errors yet. Keep this result.
+                            results.Add(completedTask.Result);
+                        }
+
+                        taskSet.Remove(completedTask);
                     }
                 }
+                finally
+                {
+                    linkedTokenSource.Cancel();
+                }
+
+                // Since we treat produce the same light bulb category for all
+                // non-error diagnostics, just use the first one.
+                return CreateResult(fullResult, results.FirstOrDefault());
```

Comment:
You only ever use results to keep around the first one.  so you could just have a single 'best' diagnostic data that you've seen.